suite/tests/client-interface/mangle_suspend.dll.c

```diff
@@ -0,0 +1,190 @@
+            if (prev_was_mov_const && val1 == val2 &&
+                val1 != 0 && /* rule out xor w/ self */
+                opnd_is_reg(instr_get_dst(inst, 0)) &&
+                opnd_get_reg(instr_get_dst(inst, 0)) == SUSPEND_TEST_REG) {
+                *user_data = (void *)val1;
+                instrlist_meta_postinsert(bb, inst, INSTR_CREATE_label(drcontext));
+            } else
+                prev_was_mov_const = true;
+        } else
+            prev_was_mov_const = false;
+    }
+#else
+        /* XXX i#3307: port to ARM if possible. */
+#endif
+    return DR_EMIT_DEFAULT;
+}
+
+static void
+suspend_func()
+{
+#ifdef X86_64
+    void **drcontexts = NULL;
+    uint num_suspended;
+    if (dr_suspend_all_other_threads(&drcontexts, &num_suspended, NULL)) {
+        if (num_suspended != 1)
+            CHECK(false, "num_suspended unexpected!");
+        dr_mcontext_t mc = { sizeof(mc), DR_MC_INTEGER | DR_MC_CONTROL };
+        if (!dr_get_mcontext(drcontexts[0], &mc))
+            CHECK(false, "dr_get_mcontext failed!");
+        /* Check xip such that we only perform the check if we are in the loop
+         * body of the test, in order to avoid races.
+         */
+        if (mc.xip >= add_instr_pc - 7 && mc.xip <= add_instr_pc + 24) {
+            if (add_instr_pc == mc.xip) {
+                if (LOOP_COUNT_REG_MC != 1) {
+                    CHECK(false, "loop count reg expected to be 1");
+                    exit(1);
+                }
+            } else if (LOOP_COUNT_REG_MC != 2) {
+                CHECK(false, "loop count reg expected to be 2");
+                exit(1);
+            }
+        }
+        int resume_count = 0;
+        while (!dr_resume_all_other_threads(drcontexts, num_suspended)) {
+            if (resume_count++ == MAX_RESUME_COUNT) {
+                CHECK(false, "resume failed!");
+                exit(1);
+            }
+        }
+    }
+#else
+    /* XXX i#3307: port to ARM if possible. */
+#endif
+}
+
+static dr_emit_flags_t
+event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst,
+                      bool for_trace, bool translating, void *user_data)
+{
+    ptr_int_t subaction = (ptr_int_t)user_data;
+
+    if (translating)
+        return DR_EMIT_DEFAULT;
+
+    if (subaction == TEST_VAL_C) {
+        if (instr_is_label(inst)) {
+#ifdef X86_64
+            byte *pc = tag;
+            instr_t *instr = instr_create(drcontext);
+            while (true) {
+                instr_reset(drcontext, instr);
+                byte *prev_pc = pc;
+                pc = decode(drcontext, pc, instr);
+                if (pc == NULL)
+                    CHECK(false, "Unexpected decode error");
+                if (instr_get_opcode(instr) == OP_add &&
+                    instr_has_rel_addr_reference(instr)) {
+                    add_instr_pc = prev_pc;
+                    instr_destroy(drcontext, instr);
+                    break;
+                }
```

Comment:
There's a mem leak if there's no break -- but I guess there's an infinite loop before that.  In non-test code there should be a counter and escape.  I guess in a test it will time out and fail so we'll know so ok.

Reply:
Obsolete, see above.